src/katgpucbf/xbgpu/engine.py

```diff
@@ -396,96 +817,34 @@ def __init__(
-            rx_item = RxQueueItem(buffer_device, present)
-            self._rx_free_item_queue.put_nowait(rx_item)
-
-        for _ in range(n_tx_items):
-            buffer_device = self.correlation.slots["out_visibilities"].allocate(allocator, bind=False)
-            saturated = self.correlation.slots["out_saturated"].allocate(allocator, bind=False)
-            present_ants = np.zeros(shape=(n_ants,), dtype=bool)
-            tx_item = TxQueueItem(buffer_device, saturated, present_ants)
-            self._tx_free_item_queue.put_nowait(tx_item)
-
-        for _ in range(n_free_chunks):
-            buf = self.correlation.slots["in_samples"].allocate_host(self.context)
-            present = np.zeros(n_ants * self.heaps_per_fengine_per_chunk, np.uint8)
-            chunk = recv.Chunk(data=buf, present=present, stream=self.receiver_stream)
-            chunk.recycle()  # Make available to the stream
-
-        # NOTE: This value staggers the send so that packets within a heap are
-        # transmitted onto the network across the entire time between dumps.
-        # Care needs to be taken to ensure that this rate is not set too high.
-        # If it is too high, the entire pipeline will stall needlessly waiting
-        # for packets to be transmitted too slowly.
-        self.dump_interval_s: float = self.timestamp_increment_per_accumulation / adc_sample_rate_hz
-
-        self.send_stream = XSend(
-            n_ants=n_ants,
-            n_channels=n_channels_total,
-            n_channels_per_stream=n_channels_per_stream,
-            dump_interval_s=self.dump_interval_s,
-            send_rate_factor=send_rate_factor,
-            channel_offset=self.channel_offset_value,  # Arbitrary for now - depends on F-Engine stream
-            context=self.context,
-            packet_payload=dst_packet_payload,
-            stream_factory=lambda stream_config, buffers: make_stream(
-                dest_ip=outputs[0].dst.host,
-                dest_port=outputs[0].dst.port,
-                interface_ip=dst_interface,
-                ttl=dst_ttl,
-                use_ibv=dst_ibv,
-                affinity=dst_affinity,
-                comp_vector=dst_comp_vector,
-                stream_config=stream_config,
-                buffers=buffers,
-            ),
-            tx_enabled=self._init_tx_enabled,
+        # it back to the corresponding _(rx/tx)_free_item_queue to ensure that
+        # all allocated buffers are in continuous circulation.
+        self._rx_free_item_queue: asyncio.Queue[RxQueueItem] = monitor.make_queue("rx_free_item_queue", n_rx_items)
+
+        rx_data_shape = (
+            heaps_per_fengine_per_chunk,
+            n_ants,
+            n_channels_per_stream,
+            n_spectra_per_heap,
+            N_POLS,
+            COMPLEX,
         )
+        for _ in range(n_rx_items):
+            # TODO: Abstract dtype, as per NGC-1000
+            buffer_device = accel.DeviceArray(context, rx_data_shape, dtype=np.int8)
```

Comment:
You've changed the way that the `buffer_device` is assigned here. Why?

Do you want to avoid reaching into one of the Pipeline objects and using the slots' `allocate_host` function?

Reply:
I'm happy with this change. Once we have B pipelines, it will be more difficult to generically access the appropriate slots within the pipelines.